### DISCUSSION TITLE
Push built files to repo using Actions

### DISCUSSION THREAD

**Discussion by Andre601** [community/community#26615] [Actions]
<p>I want to setup a GitHub Action, which would build two things - a wiki using MkDocs and Javadoc using Gradle’s Javadoc command - and then push those towards the repository itself, so that it can then be displayed using GitHub pages.</p>
<p>My current workflow is setup to build the pages and all I need now is a way to push those towards the repository itself.</p>
<p>What would be a good way to do it?</p>
<p>The Workflow can be found here:</p>
<p><a href="https://github.com/botblock/JavaBotBlockAPI/blob/master/.github/workflows/build_docs.yml" rel="nofollow noopener">https://github.com/botblock/JavaBotBlockAPI/blob/master/.github/workflows/build_docs.yml</a></p>
<p>Other things to know:</p>
<p>GitHub pages is set up to take the content from the docs directory to then display. So using the gh-pages branch and maybe MkDocs’ gh-deploy command isn’t a solution for me.</p>
<p>Additionally do I need to build the wiki with MkDocs first before the Javadoc due to MkDocs deleting everything in the target-directory (docs folder).</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/andre601">@andre601</a> ,</p>
<p>You can use the  <strong>git push</strong> command to push your local changes to the remote GitHub repository.</p>
<p>Before execute the git push, you need to set the  <strong>git config</strong>.</p>
<p>A simple demo to push changes to remote in the workflow:</p>
<pre><code>- name: update file and push to remote
        run: |
          echo "Test push commit to the remote in the workflow" &gt;&gt; test.txt

          git config --global user.name "github-actions[bot]"
          git config --global user.email "41898282+github-actions[bot]@users.noreply.github.com"

          git add -A
          git commit -m "Update test.txt"
          git push
</code></pre>
<p>Here I use the bot account <strong>github-actions[bot]</strong> to set the  <strong>git config</strong>. You also can set yourself account if you want.</p>

**Reply by yinshiyi** [reply]
I used your method got an error, do you know how to fix it? Thanks
```
fatal: You are not currently on a branch.
To push the history leading to the current (detached HEAD)
```

**Comment by Andre601** [comment]
<p>So this would only push files that are in the docs directory on the workflow?<br>
I don’t really want to test this out, as I fear it could mess up the repository, which I would like to avoid.</p>
<p>Also, does this mean that the workflow already has a git-repo setup? Doesn’t it also require it to first being setup in order to push stuff, or does GitHub (Actions) already set the workflow to have the repository as a remote set?</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/andre601">@andre601</a> ,</p>
<p>All my above steps are based on having checked out the repository from the remote to the local of the runner machine.<br>
The basic process is following:</p>
<pre><code>steps:
  - name: Checkout code
    uses: actions/checkout@v2

  - name: Modify/add files
    . . .

  - name: push changes to remote
  run: |
    git config --global user.name "user mane"
    git config --global user.email "user email"

    git add **\&lt;pathspec\&gt;** git commit -m "commit message"
    git push
</code></pre>
<p>Description:</p>
<p><strong>1)</strong> The <a href="https://github.com/actions/checkout" rel="nofollow noopener"><strong>checkout action</strong></a> can check out the repository from the remote to the local and also generate the local repository on the runner machine, it does not only download the source files to the local.</p>
<p><strong>2)</strong> In the <strong>git add</strong> command, you can specify which modified files ( <strong>&lt;pathspec&gt;</strong> ) you want to commit,<br>
for example, only commit the modified files, in the directory " <strong>docs</strong>":</p>
<pre><code>git add docs
</code></pre>
<p>If yo want to commit all the modified files, you can use the <strong>-A</strong> option.</p>
<pre><code>git add -A
</code></pre>
<p>More details, you can reference here: <a href="https://git-scm.com/docs/git-add" rel="nofollow noopener">https://git-scm.com/docs/git-add</a></p>

**Comment by Andre601** [comment]
<p>Thank you very much!</p>
<p>It works now and like I wanted it to.</p>
<p>But before I go another small question. I currently use a workflow on another repository to automatically push changes to the gh-pages branch of said repository and created a PAT (Personal Access Token) for it.</p>
<p>Is this even required? Because when I check my PATs is this one shown as “never used”.</p>
<p>I have it as secret in the repo and use it when adding a remote (I guess? I assume this isn’t really needed here?)</p>
<pre><code>git remote add gh-token "https://${{ secrets.GH_TOKEN}}@github.com/:user/:repo.git"
</code></pre>
<p>The entire workflow:</p>
<pre><code>name: Deploy Site

on:
  push:
    paths:
    - 'docs/**'
    - '**.yml'
    branches:
    - master

jobs:
  build:
    runs-on: [ubuntu-latest]
    steps:
    - uses: actions/checkout@v2
    - name: Set up Python 3.7
      uses: actions/setup-python@v1
      with:
        python-version: 3.7
    - name: Install dependencies
      run: |
        python -m pip install --upgrade pip setuptools
        python -m pip install -r requirements.txt
    - name: Deploy Files
      run: |
        git config user.name ${{ secrets.GH_USER }}
        git config user.email "${{ secrets.GH_MAIL }}"
        git remote add gh-token "https://${{ secrets.GH_TOKEN}}@github.com/:user/:repo.git"
        git fetch gh-token &amp;&amp; git fetch gh-token gh-pages:gh-pages
        python -m mkdocs gh-deploy -v --clean --remote-name gh-token
        git push gh-token gh-pages
</code></pre>

**Comment by ashblue** [comment]
<p>To add there is an automatically generated token <code>secrets.GITHUB_TOKEN</code> now.</p>
<aside class="onebox githubfolder">
  <header class="source">
      <img src="https://docs.github.com/assets/images/site/favicon.svg" class="site-icon" width="16" height="16">
      <a href="https://docs.github.com/en/actions/reference/authentication-in-a-workflow#permissions-for-the-github_token" target="_blank" rel="noopener nofollow ugc">docs.github.com</a>
  </header>
  <article class="onebox-body">
    

<h3><a href="https://docs.github.com/en/actions/reference/authentication-in-a-workflow#permissions-for-the-github_token" target="_blank" rel="noopener nofollow ugc">Authentication in a workflow - GitHub Docs</a></h3>

<p><a href="https://docs.github.com/en/actions/reference/authentication-in-a-workflow#permissions-for-the-github_token" target="_blank" rel="noopener nofollow ugc">//docs.github.com/en/actions/reference/authentication-in-a-workflow</a></p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by ttodua** [comment]
if anyone still needs, when using "checkout" action, you might need to add `persist-credentials: false`:
```
      - name: Checkout code
        uses: actions/checkout@v4
        with:
          ref: ${{ github.ref }}
          persist-credentials: false
```
